core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/AwsS3V4HttpProperties.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.signer;
+
+import static software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner.CHUNKED_ENCODING;
+import static software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner.PAYLOAD_SIGNING;
+import static software.amazon.awssdk.http.auth.aws.util.SignerConstant.STREAMING_UNSIGNED_PAYLOAD_TRAILER;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.safeParseLong;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.validatedProperty;
+
+import java.time.Clock;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.Header;
+import software.amazon.awssdk.http.auth.aws.AwsS3V4HttpSigner;
+import software.amazon.awssdk.http.auth.aws.checksum.ChecksumAlgorithm;
+import software.amazon.awssdk.http.auth.aws.signer.AwsV4HttpProperties;
+import software.amazon.awssdk.http.auth.aws.util.CredentialScope;
+import software.amazon.awssdk.http.auth.aws.util.HttpChecksumUtils;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+
+/**
+ * An extension of {@link AwsV4HttpProperties}, which provides access to more specific parameters
+ * used by {@link AwsS3V4HttpSigner}.
+ */
+@SdkProtectedApi
+interface AwsS3V4HttpProperties extends AwsV4HttpProperties {
+
+    static AwsS3V4HttpProperties create(SignRequest<?, ? extends AwsCredentialsIdentity> signRequest) {
+        boolean payloadSigning = validatedProperty(signRequest, PAYLOAD_SIGNING, false);
+        boolean chunkedEncoding = validatedProperty(signRequest, CHUNKED_ENCODING, false);
+        boolean unsignedStreamingTrailer = signRequest.request().firstMatchingHeader("x-amz-content-sha256")
+            .map(STREAMING_UNSIGNED_PAYLOAD_TRAILER::equals)
+            .orElse(false);
+        Optional<Long> maybeContentLength = safeParseLong(signRequest.request().firstMatchingHeader(Header.CONTENT_LENGTH)
+            .orElse(null));
+        Long contentLength = maybeContentLength
```

Comment:
I'm not sure, but I'm thinking through the tradeoff of unsignedStreamingTrailer and contentLength being Properties in this sub-interface v/s being methods on the Signer impl class.

I see that methods of BaseAwsV4HttpSigner take in the Properties which can include these additional ones. If not as properties, would have to pass maybe the SdkHttpRequest around instead.
And for the other ones like shouldSignPayload, shouldChunkEncode, expirationDuration (in the AwsV4PresignedHttpProperites), they could all go into AwsV4HttpProperties. Then we wouldn't need multiple sub-interfaces of AwsV4HttpProperties, removing a bunch of code.

I'm not sure which is better - maybe the sub-Properties approach is more flexible, even though more code. Just sharing these thoughts here. Moreover, these sub-Properties and sub-Signers are @SdkInternalApi, so we could refactor this way if we decided to, so I think this is non-blocking.

Reply:
There's definitely tradeoffs, I thought long and hard about "best" way, but there really isn't one for what we want. On one hand, we could handle everything by just passing around the request object (like you mention), and we just repeatedly check for presence of different headers or params, i.e. for trailing checksum, check if the "x-amz-trailer" is present, or for chunk encoding, check that the "Content-Encoding" header has the value "aws-chunked".